core/lib/zksync_core/src/consensus/fetcher.rs

```diff
@@ -0,0 +1,197 @@
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+
+    /// Periodically fetches the head of the main node
+    /// and updates `SyncState` accordingly.
+    async fn fetch_state_loop(&self, ctx: &ctx::Ctx) -> ctx::Result<()> {
+        const DELAY_INTERVAL: time::Duration = time::Duration::milliseconds(500);
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            match ctx.wait(self.client.fetch_l2_block_number()).await? {
+                Ok(head) => {
+                    self.sync_state.set_main_node_block(head);
+                    ctx.sleep(DELAY_INTERVAL).await?;
+                }
+                Err(err) => {
+                    tracing::warn!("main_node_client.fetch_l2_block_number(): {err}");
+                    ctx.sleep(RETRY_INTERVAL).await?;
+                }
+            }
+        }
+    }
+
+    /// Fetches genesis from the main node.
+    async fn fetch_genesis(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::Genesis> {
+        let genesis = ctx
+            .wait(self.client.fetch_consensus_genesis())
+            .await?
+            .context("fetch_consensus_genesis()")?
+            .context("main node is not running consensus component")?;
+        Ok(zksync_protobuf::serde::deserialize(&genesis.0).context("deserialize(genesis)")?)
+    }
+
+    /// Fetches (with retries) the given block from the main node.
+    async fn fetch_block(&self, ctx: &ctx::Ctx, n: MiniblockNumber) -> ctx::Result<FetchedBlock> {
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            let res = ctx.wait(self.client.fetch_l2_block(n, true)).await?;
+            match res {
+                Ok(Some(block)) => return Ok(block.try_into()?),
+                Ok(None) => {}
+                Err(err) if err.is_transient() => {}
+                Err(err) => {
+                    return Err(anyhow::format_err!("client.fetch_l2_block({}): {err}", n).into());
+                }
+            }
+            ctx.sleep(RETRY_INTERVAL).await?;
+        }
+    }
+
+    /// Fetches blocks from the main node in range `[cursor.next()..end)`.
+    pub(super) async fn fetch_blocks(
+        &self,
+        ctx: &ctx::Ctx,
+        cursor: &mut storage::Cursor,
+        end: Option<validator::BlockNumber>,
+    ) -> ctx::Result<()> {
+        const MAX_CONCURRENT_REQUESTS: usize = 100;
+        let done = |next| matches!(end, Some(end) if next>=end);
+        let mut next = cursor.next();
+        scope::run!(ctx, |ctx, s| async {
+            let (send, mut recv) = ctx::channel::bounded(MAX_CONCURRENT_REQUESTS);
+            s.spawn(async {
+                let send = send;
+                while !done(next) {
+                    let n = MiniblockNumber(next.0.try_into().unwrap());
+                    self.sync_state.wait_for_main_node_block(ctx, n).await?;
+                    send.send(ctx, s.spawn(self.fetch_block(ctx, n))).await?;
```

Comment:
JFYI, @tomg10 has run into problems recently with rate limiting caused by executing many requests concurrently. The issues were caused by the `MainNodeFetcher`, but it looks like they could still be present in this implementation as well – AFAIU, it starts fetching first `MAX_CONCURRENT_REQUESTS` from the main node almost at the same time, which could trigger rate limiting.

Reply:
what rate limit should I add then?

Reply:
added rate limiter